smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java

```diff
@@ -0,0 +1,213 @@
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
+     *
+     * @return upload service JID or null if not available
+     */
+    public DomainBareJid getUploadService() {
+        return uploadService;
+    }
+
+    /**
+     * Get max file size allowed by upload service.
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
+     *
+     * @return max file size in bytes or null if not available
+     */
+    public Long getMaxFileSize() {
+        return maxFileSize;
+    }
+
+    /**
+     * Request a new upload slot.
+     *
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @param filename name of file to be uploaded
+     * @param fileSize file size in bytes -- must be less or equal to {@link HttpFileUploadManager#getMaxFileSize()} (if available)
+     * @return file upload Slot in case of success
+
+     * @throws IllegalArgumentException if fileSize is greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NotConnectedException
+     * @throws SmackException.NoResponseException
+     */
+    public Slot requestSlot(String filename, long fileSize) throws InterruptedException,
+            XMPPException.XMPPErrorException, SmackException.NotConnectedException, SmackException.NoResponseException {
+        return requestSlot(filename, fileSize, null);
+    }
+
+    /**
+     * Request a new upload slot.
+     *
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @param filename name of file to be uploaded
+     * @param fileSize file size in bytes -- must be less or equal to {@link HttpFileUploadManager#getMaxFileSize()} (if available)
+     * @param contentType file content-type or null
+     * @return file upload Slot in case of success
+
+     * @throws IllegalArgumentException if fileSize is greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws SmackException.NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NoResponseException
+     */
+    public Slot requestSlot(String filename, long fileSize, String contentType) throws SmackException.NotConnectedException,
+            InterruptedException, XMPPException.XMPPErrorException, SmackException.NoResponseException {
+        if (uploadService == null) {
+            discoverUploadService();
+        }
+
+        if (maxFileSize != null) {
+            if (fileSize > maxFileSize) {
+                throw new IllegalArgumentException("File size must not be greater than " + maxFileSize);
+            }
+        }
+
+        SlotRequest slotRequest = new SlotRequest(filename, fileSize, contentType);
+        slotRequest.setTo(uploadService);
```

Comment:
That assumes that `uploadService` was set by `discoverUploadService()`, which may not be true. Again the `uploadServiceDiscoveryed` flag should be used and a `SmackException` should be thrown is its false and if the user did not manually specify an upload service.